### PR TITLE
STACK-1063 Implement VE and VLAN for Hardware Devices (DHCP only)

### PR DESCRIPTION
**Pull request by afrin-chakure-a10**
## Description
- Added feature to enable communication over vlans with traffic isolation for Hardware devices.

## Jira Ticket
[STACK-1063](https://a10networks.atlassian.net/browse/STACK-1063)

## Technical Approach
- Added Config option - `network_type` for user to specify `vlan` or `flat` hierarchy
- Under `[hardware_thunder]`, user can add `interface_vlan_map` to specify configuring ve via dhcp or not. Added column in `vthunders` table and `VThunder` model.
- Added validations for `interface_vlan_map` in `utils.py`
- Separated flat and vlan subflows.
- Added method in `utils` to get network driver.
- Added task to create and delete TagEthernet for both loadbalancer and member.
- Added methods to create, get and delete neutron ports.
- Refactored `config_types.py` 

## Config Changes
<pre>
<b>[a10_global]</b>
network_type = 'vlan'

<b>[hardware_thunder]</b>
devices=[
 {
 "project_id": "fake_uuid",
 "username" : "username",
 "password" : "password",
 "ip_address" : "10.43.12.137",
 "device_name" : "rack_1",
 "interface_vlan_map":{
                 {"1":
                       {"11": {"use_dhcp": True}, 
                         "12":{"use_dhcp": True}
                        }
                 }
       }
 }]
</pre>

## Test Cases
- Updated UTs to handle positive and negative test cases.

## Manual Testing
1) Add 1 VIP and 2 Members in 3 different VLANs and send traffic
2) Delete the Members and VIP one by one and check the VLANs and VE deleted
3) Add VIP and Member from same VLAN and another member in different VLAN send traffic
4) After above case Delete the Member on VIP VLAN and send traffic
5) Change the VE IPs in a10-octavia.conf, restart a10-controller worker and add a member. Check the VE IPs changed accordingly in thunder. Send traffic.
6) Again change VE IPs in conf, restart controller worker and disable a member. Check the VE IPs changed accordingly in thunder. Send traffic.
7) With network_type flat add 1 VIP and 2 Members and send traffic.
8) With above setup remove 1 Member and send traffic.
9) With above setup remove the remaining entities and check configuration.

### CODE REVIEW

**Comment by afrin-chakure-a10** [comment]
worked on latest review comments from 
https://github.com/a10networks/a10-octavia/pull/84

File: a10_octavia/controller/worker/tasks/utils.py [hunk 1]
```diff
@@ -60,3 +64,11 @@ def meta(lbaas_obj, key, default):
     except Exception:
         return default
     return meta_json.get(key, default)
+
+
+def get_a10_network_driver():
```
Comment (OmkarTelee-A10) [inline a10_octavia/controller/worker/tasks/utils.py:69]: I think these changes are already in master now

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 2]
```diff
@@ -359,3 +360,223 @@ def revert(self, vthunder, *args, **kwargs):
         except Exception as e:
             LOG.exception("Failed to revert partition create : %s", str(e))
             raise
+
+
+class TagEthernetBaseTask(task.Task):
```
Comment (OmkarTelee-A10) [inline a10_octavia/controller/worker/tasks/vthunder_tasks.py:365]: This code belongs to a10_network_task. We already have loaded network driver and all with VRID_FLOATING_IP changes, we will also have axapi decorator
All changes I see are related to networks.

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py [hunk 3]
```diff
@@ -134,14 +135,14 @@ def get_plugged_parent_port(self, vip):
     def create_port(self, network_id, subnet_id=None, fixed_ip=None):
         new_port = None
         if not subnet_id:
-            subnet_id = self.neutron_client.get_network(network_id).subnets[0]
-        fixed_ip = {'subnet_id': subnet_id}
+            subnet_id = self.get_network(network_id).subnets[0]
+        fixed_ip_dict = {'subnet_id': subnet_id}
```
Comment (OmkarTelee-A10) [inline a10_octavia/network/drivers/neutron/a10_octavia_neutron.py:139]: resolved this error in VRID code

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 4]
```diff
@@ -359,3 +360,223 @@ def revert(self, vthunder, *args, **kwargs):
...
+            self.release_ve_ip_from_neutron(vlan_id, vlan_subnet_id_dict[vlan_id])
+
+        if ve_info is None or 'use_dhcp' in ve_info and ve_info['use_dhcp']:
+            self.axapi_client.interface.ve.update(vlan_id, dhcp=True, enable=True)
+        else:
+            patched_ip = self._get_patched_ve_ip(ve_info['ve_ip_address'])
+            self.axapi_client.interface.ve.update(vlan_id, ip_address=patched_ip,
+                                                  ip_netmask=self._subnet_mask, enable=True)
+        self.reserve_ve_ip_with_neutron(vlan_id, vlan_subnet_id_dict[vlan_id])
+        return
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/tasks/vthunder_tasks.py:466]: Remove this return

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py (hunk 3 above)
Comment (hthompson-a10) [inline a10_octavia/network/drivers/neutron/a10_octavia_neutron.py:139]: Let's change this to `fixed_ip_subnet`

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py [hunk 5]
```diff
@@ -152,9 +153,48 @@ def create_port(self, network_id, subnet_id=None, fixed_ip=None):
             raise exceptions.PortCreationFailedException(message)
         return new_port
 
+    def get_port_id(self, ip):
```
Comment (hthompson-a10) [inline a10_octavia/network/drivers/neutron/a10_octavia_neutron.py:156]: let's rename to `get_port_id_from_ip`

File: a10_octavia/tests/unit/controller/worker/tasks/test_vthunder_tasks.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        mock_task._subnet_ip = SUBNET_MASK[0]
+        mock_task._subnet_mask = SUBNET_MASK[1]
+        mock_task._subnet = mock.Mock()
+        mock_task._subnet.cidr = SUBNET_MASK[2]
+        mock_task._subnet = mock.Mock()
+        mock_task._subnet.id = mock.Mock()
+        mock_task._subnet.network_id = "mock-network-1"
+        mock_task._network_driver = a10_octavia_neutron.A10OctaviaNeutronDriver()
+
+    def test_tag_ethernet_for_lb(self):
```
Comment (hthompson-a10) [inline a10_octavia/tests/unit/controller/worker/tasks/test_vthunder_tasks.py:110]: We need to change these names to `test_ClassName/function_name_operation/expteced_outcome_input`

File: a10_octavia/controller/worker/tasks/utils.py (hunk 1 above)
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/tasks/utils.py]: Done.

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py (hunk 3 above)
Comment (ssrinivasan-a10) [reply a10_octavia/network/drivers/neutron/a10_octavia_neutron.py]: Did the same as in VRID code.

File: a10_octavia/common/data_models.py [hunk 7]
```diff
@@ -138,7 +138,7 @@ def __init__(self, id=None, vthunder_id=None, amphora_id=None,
                  loadbalancer_id=None, project_id=None, compute_id=None,
                  topology="STANDALONE", role="MASTER", last_udp_update=None, status="ACTIVE",
                  created_at=datetime.utcnow(), updated_at=datetime.utcnow(), partition_name=None,
-                 hierarchical_multitenancy=None):
+                 hierarchical_multitenancy=None, interface_vlan_map=None):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/common/data_models.py:141]: Was able to remove interface_vlan_map from DB. It is just there in models and things work fine.

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 8]
```diff
@@ -325,6 +329,10 @@ def get_post_lb_rack_vthunder_association_flow(self, prefix, topology,
 
         post_create_lb_flow.add(database_tasks.UpdateLoadbalancerInDB(
             requires=[constants.LOADBALANCER, constants.UPDATE_DICT]))
+        if CONF.a10_global.network_type == 'vlan':
+            post_create_lb_flow.add(vthunder_tasks.TagEthernetForLB(
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:337]: Done.

File: a10_octavia/controller/worker/flows/a10_member_flows.py [hunk 9]
```diff
@@ -267,6 +274,10 @@ def get_rack_vthunder_create_member_flow(self):
         create_member_flow.add(a10_database_tasks.GetVThunderByLoadBalancer(
             requires=constants.LOADBALANCER,
             provides=a10constants.VTHUNDER))
+        if CONF.a10_global.network_type == 'vlan':
+            create_member_flow.add(vthunder_tasks.TagEthernetForMember(
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/flows/a10_member_flows.py:281]: Done.

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 10]
```diff
@@ -359,3 +360,223 @@ def revert(self, vthunder, *args, **kwargs):
...
+        self._subnet_ip = None
+        self._subnet_mask = None
+
+    @property
+    def network_driver(self):
+        if self._network_driver is None:
+            self._network_driver = a10_task_utils.get_a10_network_driver()
+        return self._network_driver
+
+    def reserve_ve_ip_with_neutron(self, vlan_id, subnet_id):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py:386]: Done.

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 11]
```diff
@@ -359,3 +360,223 @@ def revert(self, vthunder, *args, **kwargs):
...
+        if not a10_utils.check_ip_in_subnet_range(ve_ip, self._subnet_ip, self._subnet_mask):
+            LOG.warning("Not creating neutron port, VE IP %s is out of range, subnet_ip %s,"
+                        "subnet_mask %s, subnet_id %s", ve_ip, self._subnet_ip,
+                        self._subnet_mask, subnet_id)
+            return
+
+        self.network_driver.create_port(self._subnet.network_id, fixed_ip=ve_ip)
+        return
+
+    def release_ve_ip_from_neutron(self, vlan_id, subnet_id):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py:399]: Done.

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py (hunk 4 above)
Comment (ssrinivasan-a10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py]: Done.

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py (hunk 3 above)
Comment (ssrinivasan-a10) [reply a10_octavia/network/drivers/neutron/a10_octavia_neutron.py]: Removed this variable itself as in VRID code.

File: a10_octavia/network/drivers/neutron/a10_octavia_neutron.py (hunk 5 above)
Comment (ssrinivasan-a10) [reply a10_octavia/network/drivers/neutron/a10_octavia_neutron.py]: Done.

File: a10_octavia/tests/unit/controller/worker/tasks/test_vthunder_tasks.py (hunk 6 above)
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_vthunder_tasks.py]: Done.

**Pull request merged**